frontend/catalyst/jax_tracer.py

```diff
@@ -14,637 +14,902 @@
-            result_shardings=result_shardings,
-        )
+    return closed_jaxpr, abstract_results
 
-        for op in ctx.module.body.operations:
-            func_name = str(op.name)
-            is_entry_point = func_name.startswith('"jit_')
-            if is_entry_point:
-                continue
-            op.attributes["llvm.linkage"] = ir.Attribute.parse("#llvm.linkage<internal>")
 
-    return ctx.module, ctx.context
+class QFunc:
+    """A device specific quantum function.
+
+    Args:
+        qfunc (Callable): the quantum function
+        shots (int): How many times the circuit should be evaluated (or sampled) to estimate
+            the expectation values
+        device (a derived class from QubitDevice): a device specification which determines
+            the valid gate set for the quantum function
+    """
+
+    # The set of supported devices at runtime
+    RUNTIME_DEVICES = (
+        "lightning.qubit",
+        "lightning.kokkos",
+        "braket.aws.qubit",
+        "braket.local.qubit",
+    )
+
+    def __init__(self, fn, device):
+        self.func = fn
+        self.device = device
+        update_wrapper(self, fn)
+
+    def __call__(self, *args, **kwargs):
+        if isinstance(self, qml.QNode):
+            if self.device.short_name not in QFunc.RUNTIME_DEVICES:
+                raise CompileError(
+                    f"The {self.device.short_name} device is not "
+                    "supported for compilation at the moment."
+                )
+
+            backend_kwargs = {}
+            if hasattr(self.device, "shots"):
+                backend_kwargs["shots"] = self.device.shots if self.device.shots else 0
+            if self.device.short_name == "braket.local.qubit":  # pragma: no cover
+                backend_kwargs["backend"] = self.device._device._delegate.DEVICE_ID
+            elif self.device.short_name == "braket.aws.qubit":  # pragma: no cover
+                backend_kwargs["device_arn"] = self.device._device._arn
+                if self.device._s3_folder:
+                    backend_kwargs["s3_destination_folder"] = str(self.device._s3_folder)
+
+            device = QJITDevice(
+                self.device.shots, self.device.wires, self.device.short_name, backend_kwargs
+            )
+        else:
+            # Allow QFunc to still be used by itself for internal testing.
+            device = self.device
+
+        with EvaluationContext(EvaluationMode.QUANTUM_COMPILATION):
+            jaxpr, shape = trace_quantum_function(self.func, device, args, kwargs)
+
+        retval_tree = tree_structure(shape)
+
+        def _eval_jaxpr(*args):
+            return jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, *args)
+
+        args_data, _ = tree_flatten(args)
+
+        wrapped = wrap_init(_eval_jaxpr)
+        retval = func_p.bind(wrapped, *args_data, fn=self)
+
+        return tree_unflatten(retval_tree, retval)
+
+
+class QJITDevice(qml.QubitDevice):
```

Comment:
Could you move this class back to pennylane_extensions? I don't think it depends on any tracer functionality :)

Reply:
Done